### PR TITLE
dns-over-https: init at 2.2.2

### PR DESCRIPTION
**Pull request by dwoffinden**
https://github.com/m13253/dns-over-https

This is a Go implementation of a DoH client and server.

Includes a nixos module with rudimentary customisation and tests.

###### Motivation for this change

I couldn't find another server in nixpkgs that allows proxying dns-over-https requests, nor a package for this one.

I wanted the ability to host my own DoH server that applications supporting it directly (e.g. Chrome/Firefox) can be configured to talk to. In my case I have nginx in front of it, and point it at a local unbound.

The package also supports a DoH client, which I haven't tested as thoroughly. unbound natively supports DoT upstreams, and there are other packages like stubby, https-dns-proxy and dnscrypt-proxy2 competing in this space.

Actually, it looks like dnsproxy may do this and more, but I haven't tried it, and it doesn't look to have a service? (https://github.com/AdguardTeam/dnsproxy https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/networking/dnsproxy/default.nix)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -2371,6 +2371,16 @@
...
     name = "Luc Chabassier";
   };
+  dwoffinden = {
+    email = "daw@hey.com";
+    github = "dwoffinden";
+    githubId = 1432131;
+    name = "Daniel Woffinden";
+    keys = [{
+      longkeyid = "rsa4096/0xF3EA503B360FBD40";
+      fingerprint = "46FC 889E BC38 100E 51E8  3245 F3EA 503B 360F BD40";
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:2381]: Please follow the contributions guide and put this into a separate commit.

File: nixos/modules/services/networking/dns-over-https.nix [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        "connectivitycheck.gstatic.com",
+        "detectportal.firefox.com",
+        "msftconnecttest.com",
+        "nmcheck.gnome.org",
+        "pool.ntp.org",
+        "time.apple.com",
+        "time.asia.apple.com",
+        "time.euro.apple.com",
+        "time.nist.gov",
+        "time.windows.com",
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/dns-over-https.nix:35]: Would you want to expand those? E.g. ubuntu.pool.ntp.org.

File: nixos/modules/services/networking/dns-over-https.nix [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                { url = "https://dns.google/dns-query"; }
+                # CloudFlare's resolver, bad ECS, good DNSSEC
+                { url = "https://cloudflare-dns.com/dns-query"; }
+                # CloudFlare's resolver, bad ECS, good DNSSEC
+                { url = "https://1.1.1.1/dns-query"; }
+                # DNS.SB's resolver, good ECS, good DNSSEC
+                { url = "https://doh.dns.sb/dns-query"; }
+                # Quad9's resolver, bad ECS, good DNSSEC
+                { url = "https://9.9.9.9/dns-query"; }
+                # CloudFlare's resolver for Tor, available only with Tor
```
Comment (SuperSandro2000) [inline nixos/modules/services/networking/dns-over-https.nix:129]:
```suggestion
                # CloudFlare's resolver via Tor
```

File: pkgs/tools/networking/dns-over-https/default.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  };
+  vendorSha256 = "07jnrhrn45diw7mgscsiwagsdqbh0p69dnwx5bi10v2jrc66zkjd";
+  goDeps = ./deps.nix;
+  meta = with stdenv.lib; {
+    homepage = "https://github.com/m13253/dns-over-https";
+    description = "High performance DNS over HTTPS client & server";
+    platforms = platforms.unix;
+    maintainers = with maintainers; [ dwoffinden ];
+    license = licenses.mit;
+  };
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/dns-over-https/default.nix:22]:
```suggestion
  pname = "dns-over-https";
  version = "2.2.2";

  src = fetchFromGitHub {
    owner = "m13253";
    repo = pname;
    rev = "v${version}";
    sha256 = "12xklf7k07vlqbd7xv9p4c6385h874k8g5sagpcg891lib174bad";
  };
  
  goPackagePath = "github.com/m13253/dns-over-https";

  vendorSha256 = "07jnrhrn45diw7mgscsiwagsdqbh0p69dnwx5bi10v2jrc66zkjd";

  subPackages = [ "doh-client" "doh-server" ];

  meta = with stdenv.lib; {
    homepage = "https://github.com/m13253/dns-over-https";
    description = "High performance DNS over HTTPS client & server";
    platforms = platforms.unix;
    maintainers = with maintainers; [ dwoffinden ];
    license = licenses.mit;
  };
```

Go module does not use deps.nix. Please remove that.

**Review by SuperSandro2000 — COMMENTED** [review]
It would be great if you could add some integration with the unbound module or at least mention it in the example values.

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104530` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>dns-over-https</li>
  </ul>
</details>

**Comment by ajs124** [comment]
As a note, knot-resolver (cc @vcunat) can do DoH. Which IMO is a protocol that really didn't need to be invented, but that's besides the point.

**Comment by vcunat** [comment]
OK, I'll add just a few notes/comments.

Unbound also supports _serving_ DoH directly now (maybe not nicely in Unbound's NixOS module yet), and perhaps dnsdist is of note.  They don't support the client side (at least Knot Resolver and Unbound)... there I strongly prefer DoT anyway.  For me the only real reason to care for DoH is that some end-clients (browsers notably) decided that they always need the additional HTTP layer in there.

> In my case I have nginx in front of it, and point it at a local unbound.

BTW, the chain _Unbound - DoH proxy - nginx_ feels a bit too long for my personal taste.  And encrypting communication _inside_ localhost doesn't really feel useful, though I understand that sometimes the clients can get some additional information more easily if they bypass the usual OS APIs (getaddrinfo).

On the whole I'm not opposed to this package, though I personally am not interested in it at all.  Perhaps it's just a bit weird to call a package exactly like the protocol itself, especially as DoH seems to be a highly competitive space.

File: nixos/modules/services/networking/dns-over-https.nix [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+          options = {
+            enable = mkEnableOption "DNS-over-HTTPS Client";
+            bootstrap = mkOption {
+              description = "Bootstrap DNS server to resolve the address of the upstream resolver";
+              type = types.listOf types.str;
+              default = [ "8.8.8.8:53" "8.8.4.4:53" ];
+              example = [
+                # Google's resolver, bad ECS, good DNSSEC
+                "8.8.8.8:53"
+                "8.8.4.4:53"
```
Comment (vcunat) [inline nixos/modules/services/networking/dns-over-https.nix:79]: Bad ECS here?  That sounds wrong... or it needs more explanation.  I'm fairly confident that this pair of Google addresses has good [client subnet](https://www.rfc-editor.org/rfc/rfc7871.html) support over port 53.

Comment (dwoffinden) [reply nixos/modules/services/networking/dns-over-https.nix:79]: Yeah, the example config upstream seems to contradict itself, likely a mistake:

https://github.com/m13253/dns-over-https/blob/b74220718f8ed0e09f2f2f09205039af556845e1/doh-client/doh-client.conf#L21 vs https://github.com/m13253/dns-over-https/blob/b74220718f8ed0e09f2f2f09205039af556845e1/doh-client/doh-client.conf#L67

File: nixos/modules/services/networking/dns-over-https.nix (hunk 3 above)
Comment (dwoffinden) [reply nixos/modules/services/networking/dns-over-https.nix:129]: Thanks, was copied from the upstream example

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (dwoffinden) [reply maintainers/maintainer-list.nix:2381]: Done, but I don't recall seeing that guide, would you mind pointing it out? I tried to find as much info as I could :blush:

File: nixos/modules/services/networking/dns-over-https.nix (hunk 2 above)
Comment (dwoffinden) [reply nixos/modules/services/networking/dns-over-https.nix:35]: Made it an option

**Comment by dwoffinden** [comment]
> As a note, knot-resolver (cc @vcunat) can do DoH. Which IMO is a protocol that really didn't need to be invented, but that's besides the point.

Thanks. knot-resolver does seem pretty powerful. And don't get me wrong, I'm not DoH's biggest fan either :)

> Unbound also supports serving DoH directly now (maybe not nicely in Unbound's NixOS module yet), and perhaps dnsdist is of note.

Nice, thanks, I didn't see that. If I can use Unbound directly, that'd remove my need for this package, and perhaps the PR should be withdrawn if nobody else needs it.

> BTW, the chain Unbound - DoH proxy - nginx feels a bit too long for my personal taste. And encrypting communication inside localhost doesn't really feel useful

In this case the nginx->doh localhost hop is unencrypted http, with nginx handling the TLS. This is actually the recommended setup: https://github.com/m13253/dns-over-https/blob/4f46b89febf08ad906be51765b1703cb61e2f90a/doh-server/doh-server.conf#L16

>  Perhaps it's just a bit weird to call a package exactly like the protocol itself, especially as DoH seems to be a highly competitive space.

Yep, I agree 100%, I was just following the upstream repo. I actually also found https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/dns/doh-proxy/default.nix and https://github.com/jedisct1/doh-server (no nixpkg), the SEO is really terrible in this space :)

**Comment by dwoffinden** [comment]
> It would be great if you could add some integration with the unbound module or at least mention it in the example values.

Added it as an example to that option. IMO there's nothing special about a local unbound, it could equally well delegate to a local bind, or another kind of proxy. I could add an option to use the local one, or delegate to resolvconf, but I think the one-line example is easiest and most flexible for now?

**Comment by vcunat** [comment]
> In this case the nginx->doh localhost hop is unencrypted http, with nginx handling the TLS.

OK, good for some savings... but I targeted the note at the whole concept of using DoH vs. the OS DNS API calls (in case both get ended at a localhost's Unbound).

**Comment by vcunat** [comment]
> > As a note, knot-resolver (cc @vcunat) can do DoH. Which IMO is a protocol that really didn't need to be invented, but that's besides the point.
> 
> Thanks. knot-resolver does seem pretty powerful.

Thanks :-)  BTW, if you want to try `services.kresd.listenDoH` etc. you can reach out to me for basic free support, thanks to the coincidence of me being both full-time upstream developer and also being involved with NixOS a lot.

**Comment by dwoffinden** [comment]
Thanks all. Had a chance to play with kresd and it should work for my needs, so closing this as there wasn't much enthusiasm :sweat_smile:
